Documentation/proposals/202405-agent-daemonset.md

```diff
@@ -0,0 +1,133 @@
+
+### 6.2. Node detecting:
+
+As pointed out in [Danny from GMP’s talk](https://www.youtube.com/watch?v=yk2aaAyxgKw), to make Prometheus Agent DaemonSet know which node it’s on, we can use [Kubernetes’ downward API](https://kubernetes.io/docs/tasks/inject-data-application/environment-variable-expose-pod-information/). In `config-reloader` container, we can mount the node name as an environment variable like this:
+```
+containers:
+- name: config-reloader
+  env:
+  - name: NODE_NAME
+    valueFrom:
+      fieldRef:
+        apiVersion: v1
+        fieldPath: spec.nodeName
+```
+
+### 6.3. Targets filtering for pods (PodMonitor support):
+
+To filter targets, Danny’s talk has pointed out two options.
+The first option is to use relabel config. For example, to filter pods we could generate:
+```
+relabel_configs:
+- source_labels: [__meta_kubernetes_pod_node_name]
+  separator: ;
+  regex: $NODE_NAME
+  replacement: $1
+  action: keep
+```
+This could be considered as “last-mile filtering”. In other words, we watch for all the changes, and then filter the changes that match.
+
+The second option is to use field selector. For example, to select pods on a specific node, we can do:
+```
+kubernetes_sd_configs:
+- role: pod
+  selectors:
+  - role: pod
+    field: spec.nodeName=$NODE_NAME
+```
+This option performs better, because we filter targets right at discovery time, and also because Kubernetes API server watch cache indexes pods by node name (as we can see in [Kubernetes codebase](https://github.com/kubernetes/kubernetes/blob/v1.30.0-rc.0/pkg/registry/core/pod/storage/storage.go#L91)). So we’ll go with this option.
+
+
+## 7. Action Plan
+
+For the implementation, we’ll do what we detailed in the How section. The common logics between StatefulSet and DaemonSet modes will be extracted into one place. We will have a separate `daemonset.go` for the separate logic of the DaemonSet mode.
+
+For the test, we will have unit tests covering new logic, and integration tests covering the basic user cases, which are:
+* Prometheus Agent DaemonSet is created/deleted successfully.
+* Prometheus Agent DaemonSet is installed on the right nodes.
+* Prometheus Agent DaemonSet scales up/down following the scale of nodes.
+* Prometheus Agent DaemonSet selects correctly the pods on the same node.
+Currently we only set up a Kind cluster of one node for integration tests. Since the test cases for DaemonSet deployment requires at least two nodes, we will need to modify the Kind cluster config for that.
+
+We’ll also need a new user guide explaining how to use this new mode.
+
+
+## 8. Follow-ups
+
+After the Goals of this proposal have been met, we’ll handle what’s stated in the Non-goals section.
```

Comment:
I'm not sure if we'll be able to solve them, to be honest. We might be over-confident when saying that we'll solve all non-goals 😅 

What do you think about saying that different design docs will be written if we find ways to solve the non-goals?

Reply:
Yeah, I'm not saying that we'll "solve" the Non-goals. I say we'll "handle what’s stated" in Non-goals. In Non-goals I stated "The non-goals are the features that are not easy to implement and require more investigation. We will need to investigate whether there are actual user needs for them, if yes, then how to best implement them". No promises here that we'll solve (all) them.

I know these non-goals are trickier and requires more investigation. But do you think it's too pessimistic to think we'll never solve them? ^^ We may not solve them in this year GSoC but we could do it later, e.g. in next year GSoC right?

> What do you think about saying that different design docs will be written if we find ways to solve the non-goals?
I thought this's obvious. They're non-goals, so clearly this proposal is not for them. We even stated them in Follow-ups. But I'll make it more clear for the readers then.